### DISCUSSION TITLE
New feature for Queues TTL: preventing deletion of a queue when it has message buildup

### DISCUSSION THREAD

**Discussion by paulwanggz** [rabbitmq/rabbitmq-server#10202] [Questions] [answered]
### Is your feature request related to a problem? Please describe.

If you set a TTL for a queue, even if there are still messages in the queue after the application (consumer) fails, the queue will be deleted when the TTL-defined "dead time" arrives.
If the RabbitMQ administrator or developer does not handle the failed consumer in time, it is very dangerous that the unprocessed messages will be deleted, and we may lose these messages permanently.

### Describe the solution you'd like

It is recommended to add a boolean parameter to define the TTL, the default is false, if it is set to true, when there are messages stacked in the queue, even if there is no consumer connection, the queue will not be deleted after the TTL time has expired.

**Comment by lukebakken** [comment, marked as answer]
My understanding is that [Queue TTL](https://www.rabbitmq.com/ttl.html#queue-ttl) is generally used to "clean up" queues that are no longer needed, and contain messages that can be discarded.

cc @michaelklishin

**Comment by michaelklishin** [comment]
TTL (mostly message TTL) and auto-deletion features give their users and RabbitMQ developers enough reasons to waste time investigating what happened as they are. We do not plan on adding more conditions.

For example, this (very reasonable if you ask me) condition is a never ending source of questions and opinions about whether TTL should delete such queues or not:

> Queues will expire after a period of time only when they are not used (a queue is used if it has online consumers)

And you are asking us to multiply the number of scenarios? No, thanks.

If you can't afford a queue to be deleted by TTL, don't use queue TTL.
